src/main/java/org/dependencytrack/tasks/metrics/VulnerabilityMetricsUpdateTask.java

```diff
@@ -59,104 +62,75 @@ private void updateMetrics() throws Exception {
-        try (final Query<Vulnerability> query = pm.newQuery(Vulnerability.class)) {
-            if (lastId != null) {
-                query.setFilter("id < :lastId");
-                query.setParameters(lastId);
-            }
-            query.setOrdering("id DESC");
-            query.range(0, 500);
-            query.getFetchPlan().setGroup(Vulnerability.FetchGroup.METRICS_UPDATE.name());
-            return List.copyOf(query.executeList());
+            /**
+             * 
+             * The created field has priotiy over the published field, which is used as a fallback
+             * However, the created field is always empty (in my instances)
+             * But we leave this mechanism and field juggling in place for backwards compatibility,
+             * and for (future) analyzers/sources that might provide this field.
+             * 
+             * BTW the queries to get these vulnerability counts are very fast,
+             * so strictly speaking there is no reason to create this extra table with metrics
+             *  
+             */
+
+            // Get metrics by published date but only if created field is null
+            Collection<YearMonthMetric> published = queryForMetrics(pm, "published", "created");
+
+            // Get metrics by created date regardless of value of published field
+            Collection<YearMonthMetric> created = queryForMetrics(pm, "created", null);
+
+            // Merge flat lists
+            published.addAll(created);
+
+            // Collect into nested map so we can sum the counts
+            Map<Object, Map<Object, Long>> metrics = published.stream()
+            .filter(ymm -> ymm.year != null)
+            .collect(Collectors.groupingBy(ymm -> ymm.year, 
+                            Collectors.groupingBy(ymm -> ymm.month, Collectors.summingLong(ymm -> ymm.count))));         
+
+            // Flatten again, but now into VulnerabilityMetrics that can be persisted
+            Stream<VulnerabilityMetrics> monthlyStream = metrics.entrySet().stream()
+            .flatMap(e -> e.getValue().entrySet().stream().flatMap(
+                v -> Stream.of(new VulnerabilityMetrics((Integer)e.getKey(), (Integer)v.getKey(), v.getValue().intValue(), measuredAt))));
+
+            // Flatten another time, for the yearly counts
+            Stream<VulnerabilityMetrics> yearlyStream = metrics.entrySet().stream()
+            .flatMap(e -> Stream.of(new VulnerabilityMetrics((Integer)e.getKey(), null, e.getValue().values().stream().mapToInt(d->d.intValue()).sum(), measuredAt)));
+
+            // Get Mutable List to merge again
+            List<VulnerabilityMetrics> vulnerabilityMetrics = new ArrayList<VulnerabilityMetrics>(monthlyStream.toList());
+            vulnerabilityMetrics.addAll(new ArrayList<>(yearlyStream.toList()));
+
+            qm.synchronizeVulnerabilityMetrics(vulnerabilityMetrics);
```

Comment:
Can be simplified to:

```suggestion
            qm.synchronizeVulnerabilityMetrics(Stream.concat(monthlyStream, yearlyStream).toList());
```

Reply:
I had that at first, but it complained about the stream already having been operated upon. But with the current version of the code it seems to work.